torch/lib/c10d/NCCLUtils.cpp

```diff
@@ -31,4 +31,76 @@ std::string ncclGetErrorWithVersion(ncclResult_t error) {
       getNcclVersion();
 }
 
+#if NCCL_VERSION_MIN(2, 7, 0)
+ncclResult_t gpuAlltoall(
+    void* sendbuff,
+    void* recvbuff,
+    size_t count,
+    size_t elem_size,
+    ncclDataType_t type,
+    ncclComm_t comm,
+    cudaStream_t stream) {
+  int numRanks;
+  int myRank;
+  C10D_NCCL_CHECK(ncclCommCount(comm, &numRanks));
+  C10D_NCCL_CHECK(ncclCommUserRank(comm, &myRank));
+  size_t rankOffset = count * elem_size;
+  for (int r = 0; r < numRanks; r++) {
+    int sendRank = (myRank + r) % numRanks;
+    int recvRank = (myRank - r + numRanks) % numRanks;
+    C10D_NCCL_CHECK(ncclSend(
+        ((char*)sendbuff) + sendRank * rankOffset,
+        count,
+        type,
+        sendRank,
+        comm,
+        stream));
+    C10D_NCCL_CHECK(ncclRecv(
+        ((char*)recvbuff) + recvRank * rankOffset,
+        count,
+        type,
+        recvRank,
+        comm,
+        stream));
```

Comment:
Hi @srinivas212, according to [nccl documentation](https://docs.nvidia.com/deeplearning/nccl/user-guide/docs/usage/p2p.html#point-to-point-communication) perhaps these ncclSend and ncclRecv calls must be surrounded with ncclGroupStart() ncclGroupEnd() to avoid deadlock

Reply:
Hi @Sergei-Lebedev - you are right. I'm relying on nccl_group_guard in ProcessGroupNCCL.cpp to ensure send/recv gets called within group semantics. 

https://github.com/srinivas212/pytorch/blob/8f78ccbe54af2b13a8b20a0016d7faf907a5253d/torch/lib/c10d/ProcessGroupNCCL.cpp#L740

For sake of clarity, I can add an explicit group start / group end. Nested group start/end does not affect performance AFAIK and after discussing with @sjeaugey we agreed this could be a good example for folks looking for alltoall examples as well.

Reply:
I see, thanks for clarification. One more question, since GroupStart/GroupEnd calls are outside of loop it means that nccl scheduler will process bunch of send/recv so these tricky indexes for sendRank and recvRank are not necessary here

Reply:
Yes, I have fixed the loop as well with the latest PR.